Feature: Import programming exercises

ungaralex opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
As in [ARTEMIS-471](https://jirabruegge.in.tum.de/browse/ARTEMIS-471?filter=-1) described, we want to support the import of programming exercises from all available (also older) courses. It should be possible to search for any existing exercise and import it into an existing course.

### Description
<!-- Describe your changes in detail -->
There are changes both on the client and server side. 
As for the server:

- Added a new endpoint for importing programming exercises, which is available to instructors and admins.
- Mainly modified the `ProgrammingExerciseService` allowing the import of the programming exercise itself, the Bamboo build plans and the Bitbucket repositories. 
- Primitive properties, s.a. the flags for the online code editor, or publishing the build plan just get copied
- Referenced attributes, s.a. the templates or the test cases get cloned and saved in order to generate new entities and identifiers
- Identifiers like the build plan key get generated again based on the target course and new exercise name
- New pageable search functionality allowing users to query all programming exercises and receiving only a page of the result (query by title of the exercise, or the course title; page size and page number are defined by the caller)

As for the client:
- New import button right next to the "Create new programming exercise" button
- New pop-up with a pageable table containing all available programming exercises. 
- The user can search for the title, or the course title of the exercise he wants to import.
- The user can sort the columns of the table
- After selecting an exercise, the programming-exercise-update component is reused for modifying the to be imported exercise and saving it.

About the pageable search: For every new character in the search bar and every time the user sorts the table, there is a call to the backend. This can result in a lot of requests. I.m.o. this should be no problem since the import functionality will only be used in the case of setting up a new course/exercises for that course. Moreover, by doing it this way, we only sent the relevant portion of the exercise table to the user, which might be more important in the future, when there are hundreds of old programming exercises.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Select an existing course
4. Click on "Import new programming exercise"
5. Search for the exercise you want to import and click on "Import"
6. Modify any values for the new exercise
7. Click on "Save"
8. The exercise should be imported and all references, repositories and plans should be copied. (Only the template repos and plans, no student participations) 

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![import_exc](https://user-images.githubusercontent.com/27979674/64736190-609f5000-d4ea-11e9-99d5-1dbdfca6c8cc.gif)

ungaralex commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java:
```diff
@@ -228,14 +231,24 @@ public void triggerBuild(ProgrammingExerciseParticipation participation) throws
+    @Override
+    public boolean isPlanActive(final String planId) {
+        try {
+            return getBambooClient().getPlanHelper().getPlan(planId).matches("(?s)^.*Enabled[. ]*: Yes.*$");
```
Changed the implementation to use an existing Bamboo REST endpoint

lovaiible commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BitbucketService.java:
```diff
@@ -193,6 +185,54 @@ public URL getCloneURL(String projectKey, String repositorySlug) {
+                throw e;
+            }
+        }
+        catch (Exception emAll) {
```
🐹

lovaiible commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseHintService.java:
```diff
@@ -81,4 +81,32 @@ public void delete(Long id) {
     }
+
+    /**
+     * Copies the hints of an exercise to a new target exercise without by cloning the hint objects and saving them
```
```suggestion
     * Copies the hints of an exercise to a new target exercise by cloning the hint objects and saving them
```

ungaralex replied:
Changed

ungaralex replied:
Fixed the doc

lovaiible commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java:
```diff
@@ -321,51 +310,37 @@ public URL getBuildPlanWebUrl(ProgrammingExerciseParticipation participation) {
-
-        String toPlan = toProject + "-" + name;
+    @Override
+    public String clonePlan(String templatePlan, String planKey, String planName) throws BambooException {
```
Can you add the documentation for this method?

ungaralex replied:
So there already is documentation for all overridden methods in the interface of the service. I.m.o. we don't really need to add more JavaDoc if the original documentation is enough. 
Especially since we would only duplicate the doc and might get out of sync in later iterations. This might result in developers reading through different kinds of documentation depending on what class they are looking at. What do you think about this @lovaiible ?

lovaiible replied:
Yes, that's ok :)

lovaiible commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java:
```diff
@@ -321,51 +310,37 @@ public URL getBuildPlanWebUrl(ProgrammingExerciseParticipation participation) {
-     */
-    public String enablePlan(String projectKey, String planKey) throws BambooException {
+    @Override
+    public String enablePlan(String planKey) throws BambooException {
```
The documentation is also missing here.

lovaiible commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java:
```diff
@@ -391,14 +366,13 @@ public String updatePlanRepository(String bambooProject, String bambooPlan, Stri
-            //TODO use REST API DELETE "/rest/api/latest/plan/{projectKey}-{buildKey}"
-            String message = getBambooClient().getPlanHelper().deletePlan(projectKey + "-" + planKey);
+            log.info("Delete build plan " + planKey);
+            //TODO use REST API PUT "/rest/api/latest/clone/{projectKey}-{buildKey}"
```
```suggestion
            //TODO use REST API DELETE "/rest/api/latest/clone/{projectKey}-{buildKey}"
```

ungaralex replied:
Changed

krusche commented on src/main/java/de/tum/in/www1/artemis/service/connectors/VersionControlService.java:
```diff
@@ -110,5 +112,16 @@
+     * @param username Can be null. The name of the user if the newly created repository should belong to a specific user
+     * @return A map containing various information about the copied repository, s.a. the slug or the cloneUrl of the new repo.
+     */
+    Map<String, String> copyRepository(URL baseRepositoryUrl, String targetProjectKey, String targetRepositorySlug, @Nullable String username);
```
please change the signature of this method to something like

VCSRepository copyRepository(String sourceProjectName, String sourceRepositoryName, String targetProjectName, String targetRepositoryName)

ungaralex replied:
I refactored the signature

krusche commented on src/main/java/de/tum/in/www1/artemis/service/connectors/ContinuousIntegrationService.java:
```diff
@@ -28,13 +28,14 @@
+     * @return The key of the new build plan
      */
-    String copyBuildPlan(String templateBuildPlanId, String username);
+    String clonePlan(String templatePlan, String planKey, String planName);
```
change signature to String copyPlan(String sourceProjectName, String sourceBuildPlanName, String targetProjectName, String targetBuildPlanName)

improve java doc

ungaralex replied:
Refactored it

krusche commented:
Please fix the merge conflicts

ungaralex commented:
Done

krusche commented:
Due to the hot fixes in `develop`, there are again merge conflicts.
@ungaralex please resolve them again.

Thanks!

ungaralex commented:
> Due to the hot fixes in `develop`, there are again merge conflicts.
> @ungaralex please resolve them again.
> 
> Thanks!

Done

krusche commented on src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:
```diff
@@ -66,6 +68,16 @@
+    ProgrammingExercise findOneWithEagerEverything(ProgrammingExercise exercise);
+
+    @Query("select p from ProgrammingExercise p left join fetch p.testCases left join fetch p.exerciseHints left join fetch p.templateParticipation left join fetch p.solutionParticipation where p.id = :#{#id}")
+    Optional<ProgrammingExercise> findByIdWithEagerTestCasesHintsAndBaseParticipations(Long id);
```
same comment as above, also please do not use the generic term `id` --> always use e.g. `programmingExerciseId`

ungaralex replied:
Renamed to exerciseId

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/VcsRepositoryUrl.java:
```diff
@@ -0,0 +1,29 @@
+
+    public abstract VcsRepositoryUrl withUser(final String username);
+
+    public URL getRegularUrl() {
```
getUrl()

ungaralex replied:
Renamed it

krusche commented on src/main/java/de/tum/in/www1/artemis/service/connectors/VersionControlService.java:
```diff
@@ -50,58 +50,58 @@
      * @return The clone URL
      */
-    public URL getCloneURL(String projectKey, String repositorySlug);
+    VcsRepositoryUrl getCloneURL(String projectKey, String repositorySlug);
```
getCloneRepositoryUrl

krusche commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java:
```diff
@@ -213,9 +191,11 @@ public void configureBuildPlan(ProgrammingExerciseParticipation participation) {
      */
     @Override
     public void triggerBuild(ProgrammingExerciseParticipation participation) throws HttpException {
+        if (!isBuildPlanEnabled(participation.getBuildPlanId())) {
```
we should avoid this and make sure that trigger is only invoked when the build plan is already enabled

krusche commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BambooUpdateService.java:
```diff
@@ -71,17 +71,17 @@ public String updatePlanRepository(String bambooProject, String bambooPlan, Stri
-                org.swift.bamboo.cli.objects.RemoteRepository bambooRemoteRepository = bambooClient.getRepositoryHelper().getRemoteRepository(bambooRepositoryName,
-                        bambooProject + "-" + bambooPlan, false);
+                log.debug("Update plan repository for build plan " + bambooPlan);
+                org.swift.bamboo.cli.objects.RemoteRepository bambooRemoteRepository = bambooClient.getRepositoryHelper().getRemoteRepository(bambooRepositoryName, bambooPlan,
```
please update the method signature to make sure we use a consistent naming, in particular key vs. name

ungaralex replied:
Renamed to `planKey`

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -573,4 +624,10 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
         }
     }
+
+    @GetMapping("programming-exercises/pageable")
```
```suggestion
    @GetMapping("programming-exercises")
```

ungaralex replied:
Changed the route

krusche commented on src/main/webapp/app/entities/programming-exercise/programming-exercise.route.ts:
```diff
@@ -120,6 +133,16 @@ export const programmingExercisePopupRoute: Routes = [
         outlet: 'popup',
     },
+    {
+        path: 'course/:courseId/programming-exercise-import',
```
please open a popup without a route. Look into the latest delete dialog changes

ungaralex replied:
Popup now doesn't have a route anymore

krusche commented on src/main/webapp/app/entities/programming-exercise/programming-exercise.route.ts:
```diff
@@ -77,6 +78,18 @@ export const programmingExerciseRoute: Routes = [
         canActivate: [UserRouteAccessService],
     },
+    {
+        path: 'course/:courseId/programming-exercise/:id/import/:targetCourseId',
```
'course/:courseId/programming-exercise/import/:sourceProgrammingExerciseId'

ungaralex replied:
Changed the route like suggested

krusche commented on src/main/webapp/app/entities/programming-exercise/programming-exercise.component.html:
```diff
@@ -14,6 +14,17 @@ <h4>
+            *ngIf="course && course.isAtLeastInstructor"
+            id="jh-import-entity"
+            class="btn btn-primary float-right jh-create-entity create-programming-exercise mr-1"
+            [routerLink]="['/', { outlets: { popup: ['course', course.id, 'programming-exercise-import'] } }]"
```
invoke a method in the component instead of using a router link

ungaralex replied:
Using a method, which just opens a modal now

krusche commented on src/main/webapp/app/entities/programming-exercise/services/programming-exercise.service.ts:
```diff
@@ -37,6 +37,12 @@ export class ProgrammingExerciseService {
         return this.http.put(this.resourceUrl + '/' + exerciseId + '/squash-template-commits', { responseType: 'text' });
     }
 
+    importExercise(programmingExercise: ProgrammingExercise): Observable<EntityResponseType> {
```
```suggestion
    importExercise(adaptedSourceProgrammingExercise: ProgrammingExercise): Observable<EntityResponseType> {
```

krusche replied:
add a comment what the hell you are doing here ;-)

ungaralex replied:
True :smile: I added some JSDoc here explaining the method and the parameter

ungaralex commented on src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:
```diff
@@ -66,6 +68,16 @@
+    @Query("select p from ProgrammingExercise p left join fetch p.studentParticipations left join fetch p.attachments left join fetch p.categories "
+            + "left join fetch p.templateParticipation left join fetch p.solutionParticipation left join fetch p.exampleSubmissions left join fetch p.exerciseHints "
+            + "left join fetch p.tutorParticipations left join fetch p.studentQuestions left join fetch p.testCases")
+    ProgrammingExercise findOneWithEagerEverything(ProgrammingExercise exercise);
```
Moved to testing repository

ungaralex commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseHintService.java:
```diff
@@ -81,4 +81,32 @@ public void delete(Long id) {
+
+        String patchedStatement = target.getProblemStatement();
+        for (final var idMapping : hintIdMapping.entrySet()) {
+            final var replacement = "$1" + idMapping.getValue() + "$3";
```
Added documentation

ungaralex commented on src/main/webapp/app/entities/programming-exercise/programming-exercise-update.component.html:
```diff
@@ -153,8 +156,10 @@ <h4 *ngIf="programmingExercise.id" id="jhi-programming-exercise-heading-edit" jh
 
                 <div class="form-group" id="field_problemStatement" name="problemStatement">
                     <label class="form-control-label" for="field_problemStatement" jhiTranslate="artemisApp.programmingExercise.problemStatement.title"></label>
+                    <jhi-programming-exercise-instructions *ngIf="isImport" [exercise]="programmingExercise" [participation]="programmingExercise.templateParticipation">
```
Added a infor "warning", letting the user now he can still edit the statement after the import

ungaralex commented on src/main/webapp/app/entities/programming-exercise/programming-exercise-update.component.ts:
```diff
@@ -73,8 +74,20 @@ export class ProgrammingExerciseUpdateComponent implements OnInit {
+            this.selectedProgrammingLanguageValue = this.programmingExercise.programmingLanguage;
         });
         this.activatedRoute.params.subscribe(params => {
+            if (params['targetCourseId']) {
```
Removed The parameter. I am now just using the courseId and the sourceExerciseId

ungaralex commented on src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:
```diff
@@ -411,12 +408,14 @@ public ProgrammingExerciseStudentParticipation resumeExercise(Exercise exercise,
+            final var projectKey = exercise.getProjectKey();
+            final var username = participation.getStudent().getLogin();
+            final var repoName = RepositoryType.TEMPLATE.getName();
+            final var newRepoUrl = versionControlService.get().copyRepository(projectKey, repoName, projectKey, username).withUser(username);
```
Exceptions are not catched, so the participation won't be updated. But in general, I agree with you, it would be great to rework the general exception handling in Artemis, this would be part of the task there (i.e. writing good exception handlers for our REST endpoints, thereby returning appropriate entities for different exceptions). 
In this case, there are only unchecked exceptions thrown by the copy method, so we shouldn't catch them, but as just described write an exception handler for them in a future PR.

ungaralex commented on src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:
```diff
@@ -66,6 +68,16 @@
+    ProgrammingExercise findOneWithEagerEverything(ProgrammingExercise exercise);
+
+    @Query("select p from ProgrammingExercise p left join fetch p.testCases left join fetch p.exerciseHints left join fetch p.templateParticipation left join fetch p.solutionParticipation where p.id = :#{#id}")
+    Optional<ProgrammingExercise> findByIdWithEagerTestCasesHintsAndBaseParticipations(Long id);
```
True, changed it

ungaralex commented:
Integrated all feedback. Most importantly, the import popup doesn't have a dedicated route any more and the update-exercise components route for the import itself only contains the target course and source exercise ID now.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:
```diff
@@ -772,4 +784,183 @@ public void delete(Long programmingExerciseId, boolean deleteBaseReposBuildPlans
+     * @return A wrapper object containing a list of all found exercises and the total number of pages
+     */
+    public SearchResultPageDTO<ProgrammingExercise> getAllOnPageWithSize(final PageableSearchDTO<String> search) {
+        var sorting = Sort.by(ProgrammingExercise.ProgrammingExerciseSearchColumn.valueOf(search.getSortedColumn()).getMappedColumnName());
```
I agree to Thilo that we need to take the permissions into account here. This is quite important so that instructors can only import programming exercises they have access to.

Please use the Principal to retrieve the current user with his groups and check if `exercise.course.instructorGroupName` is included. Ideally this is already performed in the database call. If this is not so easy doable, you can also do this in Java and simply remove programming exercises from the page which do not fulfill the criteria

ungaralex replied:
I am now checking whether the user's groups contain the group of an exercise and filter the result in the DB call accordingly

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:
```diff
@@ -772,4 +784,183 @@ public void delete(Long programmingExerciseId, boolean deleteBaseReposBuildPlans
+            copy.setWeight(testCase.getWeight());
+            copy.setExercise(targetExercise);
+            return copy;
+        }).collect(Collectors.toSet()));
```
I guess the problem identified by Thilo is caused here. You do not save the test case entity into the database...

ungaralex replied:
I now save the entity properly

krusche commented:
There is one additional issue. The Clone Dialog does not prepend the username of the student any more. We can solve this on the client side, if this is easier.

<img width="1018" alt="image" src="https://user-images.githubusercontent.com/744067/65582253-56b32d80-df7d-11e9-8d0c-de476cd9c0f8.png">

ungaralex commented:
Thanks @krusche and @thilo-behnke, these were some great findings!

> Well thought out and works, very cool!
> 
> Some comments from my side:
> 
>     * When using the search in the import dialog, every letter typed executes a request. This could be improved by using debounceTime from RxJS
> 
>     * get /programming-exercises does not perform a permission check based on the instructors group (you can import programming exercises from other instructors courses, not sure if as desired). Maybe there should be a filter checkbox next to the search to only show exercises from this course / your courses, or you should only be able to import exercises from courses you are the instructor in.
> 
>     * Test case settings are not copied (weight, afterDueDate).
> 
>     * EDIT: Please reuse our new `jhi-button` component that has a loading behavior built in. This way all buttons will behave the same way in Artemis. In your case the title of the button is hidden when it is loading. (e.g. https://github.com/ls1intum/Artemis/pull/826/files#diff-22f88b909f784e8bb5ebb113c8ec53c3R235)
> 
>     * EDIT: Deleting a copied exercise does not work. The delete call is pending and does not return a response. Maybe a communication issue with the VCS/CI?

1. I added a debounceTime to the search
2. There now is a permission check right in the DB call, so that you only receive exercises for which you also are an instructor. I also added an integration test for this.
3. Test case settings get now properly copied (I added the missing save call to the repo, thanks @krusche )
4. I am now using the new button component instead of the old ones.
5. Deleting now works fine on the server. Although I have to say, this also worked for me before making the changes (just locally). Regarding the changes: There still was a reference in the copied build plans to the old test repository. The plans now correctly link the forked repositories. The issues with deleting exercises (both original and imported) were gone after I pushed and deployed all changes on the test server.

So far, I tested the following things:

- Create a new exercise
- Import an exercise
- Import an exercise
- Participate in a new exercise
- Participate in an imported exercise
- Delete the old exercise and try to run the build jobs of the imported exercise (success --> Old exercise has been deleted and the jobs can still run, because they are only linked to the imported repos)
- Delete an imported exercise and submit some change --> Can still participate in the old exercise
- Delete an imported exercise and try to delete the old exercise afterwards

I will run a final simulation with 1000 participants and will let you know as soon as everything works smoothly.

ungaralex commented:
Simulation looks good so far. I found one bug, where username with special chars couldn't participate because the build plan key was not filtered properly. This is fixed now
